usr/src/uts/common/fs/zfs/vdev_disk.c

```diff
@@ -754,6 +758,36 @@ vdev_disk_io_start(zio_t *zio)
 
 			break;
 
+		case DKIOCFREE:
+			if (!zfs_trim)
+				break;
+			/*
+			 * We perform device support checks here instead of
+			 * in zio_trim_*(), as zio_trim_*() might be invoked
+			 * on a top-level vdev, whereas vdev_disk_io_start
+			 * is guaranteed to be operating a leaf disk vdev.
+			 */
+			if (vd->vdev_notrim &&
+			    spa_get_force_trim(vd->vdev_spa) !=
+			    SPA_FORCE_TRIM_ON) {
+				zio->io_error = SET_ERROR(ENOTSUP);
+				break;
+			}
+
+			/*
+			 * zio->io_private contains a dkioc_free_list_t
+			 * specifying which offsets are to be freed
+			 */
+			ASSERT(zio->io_dfl != NULL);
+			error = ldi_ioctl(dvd->vd_lh, zio->io_cmd,
+			    (uintptr_t)zio->io_dfl, FKIOCTL, kcred, NULL);
+
+			if (error == ENOTSUP || error == ENOTTY)
+				vd->vdev_notrim = B_TRUE;
```

Comment:
What do you think about moving this to `zio_vdev_io_assess()` for consistency with `DKIOCFLUSHWRITECACHE`. Like this https://github.com/zfsonlinux/zfs/pull/5925/commits/77745912220848c7cef80ce515aa4e665717ae2e#diff-bdeb26c3e7b126b07c42b8583f57a2b9R3655.